atc/exec/retry_error_step.go

```diff
@@ -48,18 +51,23 @@ func (step RetryErrorStep) Run(ctx context.Context, state RunState) error {
 }
 
 func (step RetryErrorStep) toRetry(logger lager.Logger, err error) bool {
-	switch err.(type) {
-	case transport.WorkerMissingError, transport.WorkerUnreachableError:
+	var transportErr *transport.WorkerMissingError
+	var unreachable *transport.WorkerUnreachableError
+	var netOpErr *net.OpError
+	re := regexp.MustCompile(`worker .+ disappeared`)
+	if ok := errors.As(err, &transportErr) || errors.As(err, &unreachable); ok {
+		logger.Debug("retry-error",
+			lager.Data{"err_type": reflect.TypeOf(err).String(), "err": err.Error()})
+		return true
+	} else if re.MatchString(err.Error()) {
+		logger.Debug("retry-error",
+			lager.Data{"err_type": reflect.TypeOf(err).String(), "err": err})
+		return true
+	} else if errors.Is(err, io.EOF) || errors.As(err, &netOpErr) {
```

Comment:
I think you can get rid of the `io.EOF` check now that we have the check for `*net.OpError` - in your case, the observed root error was an `io.EOF`, but the true root cause was a network issue because the worker went away.

So, just having the check for `*net.OpError` works for your case (where you were seeing `io.EOF`) and for other more general network errors, so you probably don't need both (unless there's some case where you'd observe an `io.EOF` not at the network layer - not sure what that would be, though)

Reply:
@aoldershaw agree in this case it's network level error so we will cover much more errors there. Removed this EOF and pulled latest master. Thanks again for all suggestions